File: javascript/applications/stickstick/resources/data/default.js

prudence.include('../libraries/stickstick/data/');
prudence.include('../libraries/json2/');

function handleInit(conversation) {
    conversation.addMediaTypeByName('text/plain');
    conversation.addMediaTypeByName('application/json');
}

function handleGet(conversation) {
    var form = conversation.resource.request.resourceRef.queryAsForm;
    var fresh = form.getFirstValue('fresh') == 'true';
    
    var maxTimestamp;
    var boardList = [];
    var notes;

    var connection = getConnection(fresh);
    try {
	    var boards = getBoards(connection);
	    if(boards != null) {
		    for(var i in boards) {
		    	var board = boards[i];
		        boardList.push(board.id);
		        var timestamp = board.timestamp;
		        if((maxTimestamp == null) || (timestamp > maxTimestamp)) {
		            maxTimestamp = timestamp;
		        }
		    }
	    }
	    else {
	        return null;
	    }
	
	    notes = getNotes(connection);
    }
    finally {
    	connection.close();
    }

    if(maxTimestamp != null) {
        conversation.modificationTimestamp = maxTimestamp;
    }
    return JSON.stringify({boards: boardList, notes: notes});
}

function handleGetInfo(conversation) {
    var connection = getConnection();
    try {
    	return getBoardMaxTimestamp(connection);
    }
    finally {
    	connection.close();
    }
}

function handlePut(conversation) {
    // Note: You can only "consume" the entity once, so if we want it
    // as text, and want to refer to it more than once, we should keep
    // a reference to that text.
    
    var text = conversation.entity.text;
    var note = JSON.parse(String(text));
    
    var connection = getConnection();
    try {
    	addNote(note, connection);
        updateBoardTimestamp(note, connection);
    }
    finally {
    	connection.close();
    }
    
    return handleGet(conversation);
}

File: ruby/applications/stickstick/resources/data/default.rb

require 'java'
import org.json.JSONObject

require 'stickstick/data.rb'

def handle_init conversation
    conversation.add_media_type_by_name 'text/plain'
    conversation.add_media_type_by_name 'application/json'
end

def handle_get conversation
    form = conversation.resource.request.resource_ref.query_as_form
    fresh = form.get_first_value('fresh') == 'true'
    
    max_timestamp = nil
    board_list = Array.new
    notes = nil

    connection = get_connection fresh
    begin
	    boards = get_boards connection
	    if !boards.nil?
		    for board in boards
		        board_list << board['id']
		        timestamp = board['timestamp']
		        if max_timestamp.nil? || (timestamp > max_timestamp)
		            max_timestamp = timestamp
		        end
		    end
	    else
	        return nil
	    end
	
	    notes = get_notes connection
    ensure
    	connection.close
    end

    if !max_timestamp.nil?
        conversation.modification_timestamp = max_timestamp
    end
    return JSONObject.new({'boards' => board_list, 'notes' => notes})
end

def handle_get_info conversation
    connection = get_connection
    begin
    	return java.lang.Long.new get_board_max_timestamp(connection)
    ensure
    	connection.close
    end
end

def handle_put conversation
    # Note: You can only "consume" the entity once, so if we want it
    # as text, and want to refer to it more than once, we should keep
    # a reference to that text.
    
    text = conversation.entity.text
    entity = JSONObject.new(text)
    note = {}
	for key in entity.keys
		note[key] = entity.get key
	end
    
    connection = get_connection
    begin
    	add_note note, connection
        update_board_timestamp note, connection
    ensure
    	connection.close
    end
    
    return handle_get conversation
end